### ISSUE TITLE
Clarification on how to compute forces.mean and forces.stdev for EquiformerV2 denoising config

### ISSUE THREAD

**Issue by Juban0102** [atomicarchitects/DeNS#4]
Hi, I’m preparing a custom LMDB dataset for training the EquiformerV2 (denoising_forces) model. I plan to take the equiformerv2 with denoise OC20 S2EF-2M checkpoint and fine-tune it with my own dataset.
The OC20 S2EF-2M YAML config includes the following normalization block:

transforms:
  normalizer:
    energy:
      mean: <...>
      stdev: <...>
    forces:
      mean: 0.0
      stdev: 2.8873...

I was able to compute the energy mean and stdev easily using np.mean over all energy values of my dataset.
However, I'm confused about the correct procedure for forces normalization, since the forces from my lmdb dataset are in matrix form. 

My question:
1) Should forces.mean and forces.stdev be computed as scalars over all force components (x, y, z) across the entire dataset?(i.e., flattening all x,y,z force component)

2) Also, I noticed that the default mean of the config file is 0. Is this computed from the OC20 dataset? Or is this normalized to be centered around 0?

Best regards,
Joevan

**Comment by yilunliao** [comment]
Hi @Juban0102 

Thanks for your question.
My comments are as follows:

---

> 1. Should forces.mean and forces.stdev be computed as scalars over all force components (x, y, z) across the entire dataset?(i.e., flattening all x,y,z force component)

Yes. You can use `torch.norm(forces, dim=-1)` to compute the L2 norm of forces. More details are that we typically want to initialize the output xyz components to be from N(0, 1), and therefore, we would normalize force labels by `(the mean of L2 norm of forces) / sqrt(3)` (Consider the simple case of xyz = (1, 1, 1)).  


> 2. Also, I noticed that the default mean of the config file is 0. Is this computed from the OC20 dataset? Or is this normalized to be centered around 0?

The mean of forces is always 0 since we are using equivariant networks, where vectors of degree > 0 have no bias terms.

---

Let me know if anything is unclear or if you have further questions.

**Comment by Juban0102** [comment]
Hi @yilunliao. Thanks again for the explanation. I just want to make sure I’m correctly understanding how the stdev should be computed for my dataset.

From your suggestion, my interpretation is:

1.  compute the L2 norm of the force vector for each atom,
2.  calculate `(the mean of L2 norm of forces) / sqrt(3)`,  for each frame.
3.  combine all values from (2), then calculate the stdev.

And for the mean, we just set it to 0. Could you please let me know if this is correct?

**Comment by yilunliao** [comment]
For the stdev (or more precisely, the scaling factor) of forces , it would be:
1. Compute the L2 norm of each atom,
2. Take the mean over all the atoms (not frames) and then optionally divide the mean by `sqrt(3)`.
There is no additional calculation of standard deviation after 2.

You can find some implementations here: https://github.com/atomicarchitects/nequix/blob/main/nequix/data.py#L432-L442


For the mean of forces, yes.

**Comment by Juban0102** [comment]
Thank you so much for the response! I'll close this issue now and might open a new one in the future if any new questions come up. Thanks!

Best regards, 
Joevan
